[RFC] Dynamic shared memory implementation.

lorc opened the pull request:
Hello,

I want to propose new approach for shared memory. Idea is not to use carveout in memory, but to use memory provided by Normal World. Normal World allocates memory at its side and provided list of pages to OP-TEE, so OP-TEE kernel can map them into its virtual space.

This is not the final version, because it should be merged with @jenswi-linaro MOBJs framework. But anyway, it was tested on simple tests. With this approach I was able to number allocate large shared buffers (e.g. 2M each).

There are number of patches in this PR. First patches are quite generic and I can create separate PRs to merge them if maintainers wish. Most important are two last patches: one of them adds shared memory subsystem and second one puts it into use.

There are one problem that is not resolved at this time - we need to list of shared pages to TA memory manager. I hope MOBJs will help with this.

There are corresponding kernel and client changes based on @jenswi-linaro patches for registering shared memory. I'll present them later.

lorc commented:
Checkpatch have failed two times. But one error looks like a bug in checkpatch (or I don't understand what it wants from me). Another message is about long line. But there are string constants, which should not be divided into two parts (because string constant division also triggers checkpatch error).

prime-zeng commented on core/arch/arm/include/mm/tee_mmu_defs.h:
```diff
@@ -69,7 +69,7 @@
 
 #define TEE_MMU_L2_NUM_ENTRIES		(TEE_MMU_L2_SIZE / 4)
-#define TEE_MMU_L2_SIZE			(1 << 10)
+#define TEE_MMU_L2_SIZE			(1 << 13)
```
RES_VASPACE_SIZE is 10MB, so the L2 entries is not enough here?

etienne-lms replied:
this is the size of a nonlpae L2 table. should remain 1KByte.

prime-zeng commented on core/arch/arm/mm/shmem.c:
```diff
@@ -0,0 +1,587 @@
+		unsigned int idx;
+
+		while (true) {
+			if (!core_mmu_find_table(vaddr, 99, &tbl_info)) {
```
Use UINT_MAX instead of 99.

prime-zeng commented on core/arch/arm/mm/shmem.c:
```diff
@@ -0,0 +1,587 @@
+ *   TEE_ERROR_XXX - on error
+ */
+
+int shmem_map_buffer(paddr_t pa, size_t size, paddr_t *pages,
```
The page list dynamic mapping is a common function,  can we introduce a new function like map_page_list,  and the shmem_map_buffer is a wrapper function of the map_page_list here?

jenswi-linaro commented:
Please create a new PR for the first 6 patches. They look quite easy and shouldn't take much time to review.

jenswi-linaro commented:
Please rebase this now that all prerequisites are in place.

lorc commented:
Patches was rebased on master branch.

jenswi-linaro commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -1132,6 +1150,12 @@ void *phys_to_virt(paddr_t pa, enum teecore_memtypes m)
+		va = map_pa2va(find_map_by_type_and_pa(m, pa), pa);
+		if (!va)
+			va = shmem_pa2va(pa);
+		break;
```
+1

jenswi-linaro commented on core/arch/arm/tee/entry_std.c:
```diff
@@ -327,11 +366,112 @@ static void entry_cancel(struct thread_smc_args *smc_args,
+	if (arg == NULL)
+		goto err;
+
+	num_params = arg->num_params;
```
`num_params` will be read once again in `tee_entry_std()` below after this function has returned. What if normal world succeeds in using this window to increase `num_params`?

jenswi-linaro commented on core/include/mm/tee_mm.h:
```diff
@@ -62,6 +62,9 @@ extern tee_mm_pool_t tee_mm_sec_ddr;
 extern tee_mm_pool_t tee_mm_vcore;
 
+/* Shared memory pool */
+extern tee_mm_pool_t tee_mm_shm __early_bss;
```
+1

jenswi-linaro commented on core/include/optee_msg.h:
```diff
@@ -55,6 +55,11 @@
+ * Special parameter type denoting that command buffer continues on
+ *  specified page
+ */
+#define OPTEE_MSG_ATTR_TYPE_NEXT_FRAGMENT	0xc
```
This together with `OPTEE_MSG_ATTR_NESTED` is quite complicated. To make it easier to understand please describe how they can be used in one or two example use cases.

etienne-lms commented on core/arch/arm/kernel/thread.c:
```diff
@@ -1250,22 +1252,40 @@ void thread_rpc_free_arg(uint64_t cookie)
-		thread_rpc_free_arg(co);
-		pa = 0;
-		co = 0;
+
```
If needing to call a RPC to allocate discontiguous buffer, I think RPC should rather have NonSecure allocating the buffer, registering it to OP-TEE core and then returning to RPC allocation caller with the registered buffer ID. For the RPC basic communication, default SHM is ok, no real need to complexify with pure "registered" SHM.

jenswi-linaro replied:
This would be the use case where a TA is loaded.

For normal world to call back and register an SHM, it would need yet another thread in secure world as we don't support calling back from an RPC on the same thread.

lorc commented:
@jenswi-linaro , I tried to dump all `shmem.c` code and switch to mobjs. And now I'm thinking that code in `thread.c` should be mobj-aware. E.g. `thread_rpc_alloc_payload()` should return mobj and so on. What do you think about this? 

Also, I don't know what to do with checks like `core_pbuf_is`. Are they necessary in mobj-aware code? Because mobjs provide own security checks.

etienne-lms commented:
@lorc & @jenswi-linaro, regarding `core_pbuf_is`, i think it is very important that these macros are used when creating a mobj based on a physical location provided by nonsecure world. Once a mobj is safely created, we can rely on the mobj native checks (based on their identifier/property flags rather than physical location).

For example, at core entry, nonsecure provides a buffer location (eventually a list of physical chunks) to be handled as a shm buffer. The core must verify each physical location against authorized SHM locations before creating a "mobj" for the buffer.

jenswi-linaro commented:
I agree, `thread.c` should be mobj-aware.

lorc commented:
Sorry for late answer, I was on sick leave. @etienne-lms, my idea was to put `core_pbuf_is` into mobj allocation code. In this way you will have only valid `mobj` or does not have it at all. I think, in this case extra `core_pbuf_is` check is somewhat redundant. What do you think?

I'm asking because I had removed `shmem` code, so check like this (https://github.com/OP-TEE/optee_os/pull/1232/files#diff-21c5a372f38c092a429fd2b7e8f71ef3R543) is no longer possible.  

Anyway, code in `thread.c` and `entry_std.c` looks much cleaner with mobjs used.

etienne-lms commented:
hi @lorc , hope you feel better.
> my idea was to put core_pbuf_is into mobj allocation code. In this way you will have only valid mobj or does not have it at all.

I agree, This is the best way. Get a mobj and once ok only rely on the mobj API.

With your proposal, you plan to create a mobj for each of the physically contiguous chunks of the registered shm ? I rather like the proposal from Jens to extend mobj to support physically discontiguous memory and hide this complexity behind inside the mobj layer (easy to say but I'm not sure it is less complex to implement/maintain :).

lorc commented:
Hi all,

There is a new version of registered SHM. I get rid of shmem.c and moved all code to mobjs, including code in `thread.c`, FS, sockets and so on.

jenswi-linaro commented on core/arch/arm/plat-vexpress/platform_config.h:
```diff
@@ -221,7 +221,7 @@
 
 #define DRAM0_BASE		UINTPTR_C(0x40000000)
-#define DRAM0_SIZE		(UINTPTR_C(0x40000000) - CFG_SHMEM_SIZE)
+#define DRAM0_SIZE		(UINTPTR_C(0x42100000) - CFG_SHMEM_SIZE)
```
I'd rather update `build/qemu_v8.mk`.
Long term we could read (or supplement) the DRAM config from DT on the QEMU platforms.

jenswi-linaro commented on core/arch/arm/include/sm/optee_smc.h:
```diff
@@ -93,6 +93,12 @@
+ * Privileged call flag. Used by supplicant. Asks OP-TEE to use reserved
+ * thread if all regular threads are used.
+ */
+#define OPTEE_SMC_FLAGS_PRIVILEGED	BIT(0)
```
We can't waste that much memory for this. There has to be some other way.

lorc replied:
You mean that you don't want to allocate whole register for flags? Possibly we can pass flag in r0 then.

jenswi-linaro replied:
No, I'm concerned about the stack.

lorc replied:
Ah, I see. We need 3 pages per thread for stack, right? This feature will require one extra thread, so additional 12Kb of memory will be used. Is this too much?

I can imagine another approach: save context (and stack) of a suspended thread to dynamically allocated buffer and use that thread for a privileged call. But what if we couldn't allocate buffer for context? There will be deadlock.

Another approach is to try to squeeze privileged thread stack into one of the suspended thread's stack. But this is also prone to problems with the stack size.

Probably we can forbid STD SMC calls from a RPC call. Then there will be no need to a privileged thread. But in that case we will need another call path to register SHM in a RPC request. I had that solution (maybe you remember `OPTEE_MSG_ATTR_TYPE_NEXT_NESTED`) in earlier versions of the pacthes. But there was the extra param type (`NESTED` one), additional logic, etc, etc. I didn't liked it, so I choose simpler approach with existing shared memory calls.

Maybe you have ideas on how to avoid deadlock?

jenswi-linaro replied:
One extra privileged thread isn't enough, one extra privileged thread per normal thread is actually needed to avoid deadlocks. That also lead me into the idea that if during RPC registering of SHM is needed it can be done using the suspended thread currently doing the RPC. If we only allow registration of SHM in this case I don't think we'll have a problem with stack overruns.

However, it will result in some extra ping-pong and it may be a bit complicated to find out which thread did the RPC and do a call inside that.

I'm afraid I've forgotten how `OPTEE_MSG_ATTR_TYPE_NEXT_NESTED` worked, but I think the best solution would be to return of `OPTEE_MSG_RPC_CMD_SHM_ALLOC` to carry the information needed to register the shared memory. Do you really needed `OPTEE_MSG_ATTR_TYPE_NEXT_NESTED` for that?

lorc replied:
As I can see this, one extra privileged thread is sufficient. Can you give me example of calls where this will be not enough? 

We can use suspended thread's stack to register SHM, but there will be additional code to handle this specific case. I can do this if everyone is fine with this.

`OPTEE_MSG_ATTR_TYPE_NESTED` was needed because we need space to store list of pages. RPC command buffer have limited size (2 or 4 parameters, I don't quite remember) and it is allocated at OP-TEE side, so we can't reallocate it on kernel side. `OPTEE_MSG_ATTR_TYPE_NESTED` denoted PA of first page with actual list `struct optee_msg_param`. But, this again introduced new logic: code in `thread.c` needed to map additional pages, read list of parameters, construct new mobj... I was tired with debugging this, so I ditched it in favor of reusing existing calls that already registered shared memory. Code became much smaller and simpler. 

The existing approach is smallest in codebase and simplest of all variants I have tried. But, I can miss something, so I'm open to new ideas.

jenswi-linaro replied:
I suppose it's possible with careful programming to avoid needing more than one reserved thread, but that means constantly being aware of this restriction and thus accumulating technical debt. Since this will affect the ABI it's important to get it right from the start.

But regardless wasting 8k for this is too much, especially since there's other options.

lorc replied:
As I can see it, one thread will be enough in all cases except case when privileged thread will issue own RPC. That is the biggest problem in my opinion.

Can we allocate stack dynamically with `malloc()`?

jenswi-linaro replied:
I agree that it's a big problem as it can occur when taking a mutex. :-)
To allocate the stack with `malloc()` is likely to fail due to fragmented heap (8k is a significant part of ~32k).

lorc replied:
Okay, point taken. Generally speaking, RPC handling code should not issue OP-TEE calls because this calls can lead to more RPCs.

But in this particular situation with SHM we can:

a) make exception for special `register_shm` call
b) rework the way how results are returned from RPC

Variant a) leads to alterations to tasker/scheduler code. Variant b) leads to more complex ABI between OP-TEE,  kernel driver and supplicant. I prefer b) because I did it once.

What do you think?

jenswi-linaro replied:
I prefer b) too.

lorc commented:
@jenswi-linaro I had reworked this in a way you asked. I have added a new parameter type and corresponding logic in `thread.c`
Probably, it is more clear from kernel size: https://github.com/linaro-swg/linux/pull/29 (check out `rpc.c`).

jenswi-linaro commented on core/arch/arm/kernel/misc.c:
```diff
@@ -0,0 +1,242 @@
+	if (!mobj)
+		goto err;
+
+	params = (struct optee_msg_param *)
```
Cast not needed.

jenswi-linaro commented on core/arch/arm/kernel/misc.c:
```diff
@@ -0,0 +1,242 @@
+	params = (struct optee_msg_param *)
+		mobj_get_va(mobj, pa_params & SMALL_PAGE_MASK);
+	pages_cnt = 1;
+	for (uint32_t i = 0; i < num_params; i++) {
```
Please declare local variables at the beginning of a block.

jenswi-linaro commented on core/arch/arm/kernel/misc.c:
```diff
@@ -0,0 +1,242 @@
+				goto err;
+
+			pages_cnt++;
+			params = (struct optee_msg_param *)mobj_get_va(mobj, 0);
```
Cast not needed

jenswi-linaro commented on core/arch/arm/kernel/misc.c:
```diff
@@ -0,0 +1,242 @@
+			params = (struct optee_msg_param *)mobj_get_va(mobj, 0);
+		} else {
+			params++;
+			if (((vaddr_t)params & SMALL_PAGE_MASK) == 0)
```
It wouldn't hurt with a comment stating that we're checking that we're not overflowing into the next page.

Preferred style is `if (!((vaddr_t)params & SMALL_PAGE_MASK))`

jenswi-linaro replied:
I disagree, but it's no big deal. Keep it as is if you think it's important.

jenswi-linaro commented:
Please create a new PR for "spinlock: add cpu_spin_lock_xsave()/xrestore() functions" and
"tee_mm: add locking while working with pool->entry" to get them out of the way of this review.

lorc commented:
@jenswi-linaro Thank you for review.

Regarding that two patches, there are PR for them: https://github.com/OP-TEE/optee_os/pull/1356
But looks like it is stalled....

jenswi-linaro commented:
Sorry, I've forgot about #1356. I'll look at it again.

jenswi-linaro commented:
I'm starting to have some doubts of how we store physical addresses of a non-contiguous buffer in the OP-TEE MSG protocol. For each page we use 32 bytes to pass the information. I think it was my idea from the beginning so I have only myself to blame.

etienne-lms commented:
A generic comment: this change considers registered shm memory pages as non-secure page, wherever these belong to a secure memory or a non secure memory (DRAMx_BASE/_SIZE can cover both).

A far as such physical pages will be always mapped non-secure from OP-TEE core side, it looks fine, but there is a drawback. If one registers a secure physical page as non-secure shm, OP-TEE will access the page with non-secure permission, which is right, but it will likely trig an error (as a memory firewall signal) ending in a system access violation. For security view, it is fine, nothing will leak. But from an OP-TEE development, do we expect the core to behave this way ? I would rather have the core validating the physical pages, one by one, before it accepts to register the page list as shared memory. 

What's your feeling about it?

jenswi-linaro commented:
@etienne-lms , I agree the addresses should be verified.

lorc commented:
@jenswi-linaro, If I remember well, I saw that approach with array of params in your experiments with registered memory. But we can pass just plain array of page addresses if you think that it is will be better.

@etienne-lms check out `mobj_reg_shm_alloc` in `mobj.c`. I check every page against `CORE_MEM_EXTRAM`. But you are saying that `CORE_MEM_EXTRAM` covers also secure memory? Then it is not good. Is there a way to check if page belongs to non-secure RAM?

etienne-lms commented:
> Is there a way to check if page belongs to non-secure RAM?

This is quite platform specific. And the core does not provide a generic way to register all the system secure memory.

As of today, core registers the secure memory mapped into OP-TEE core (`register_phys_mem()` with appropriate types) and the SDP specific memories (`register_sdp_mem()`) if any.

If the system integrates other secure memories that are not mapped/accessed from the core, there are currently no way to define then. Maybe we should have a `register_extra_secure_mem()` to allow platform to register them. And parsing the registers `phys_mem` + `sdp_mem` + `extra_secure_mem` tables, and insuring the SHM is in DDR (as you currently do) would insure the memory is safe to be handled as non-secure shared memory.

Or maybe, as such extra secure memories are outside of the scope of the OP-TEE, it can be ok to not consider them as far as OP-TEE will only, access them with non-secure permission, hence issuing a access violation if it attempts to access them. If so, then "inside DDR + outside registered OP-TEE secure memory + outside registered SDP memory" is enough to validate OP-TEE shm candidates.

lorc commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -691,8 +698,18 @@ int core_va2pa_helper(void *va, paddr_t *pa)
+			((map->pa + (uintptr_t)va - map->va) &
+			 ~(map->region_size - 1));
+	else
+		if (!arm_va2pa_helper(va, pa))
```
Okay, I'll try to rework this in other way.

lorc commented on core/arch/arm/mm/mobj.c:
```diff
@@ -312,6 +306,335 @@ struct mobj *mobj_mm_alloc(struct mobj *mobj_parent, size_t size,
+	return NULL;
+}
+
+struct mobj *mobj_optee_shm_from_reg(struct mobj *mobj_reg_shm)
```
It had use once, but yep, now it is not needed.

jenswi-linaro commented on core/arch/arm/include/kernel/misc.h:
```diff
@@ -28,15 +28,82 @@
 uint32_t read_mode_sp(int cpu_mode);
 uint32_t read_mode_lr(int cpu_mode);
 
+/**
```
`msg_param` could be used as both prefix and basename of the file.
`extract_pages_from_params()` would then become `msg_param_extract_pages()` and so on.

jenswi-linaro commented on core/arch/arm/include/kernel/misc.h:
```diff
@@ -28,15 +28,82 @@
+ * return:
+ *	true on success, false on failure
+ */
+bool fill_memparam(struct optee_msg_param *param, struct mobj *mobj,
```
Please use `msg_param` instead, the name of this function could then be `msg_param_fill_memparam()`. Perhaps "fill" should be replaced by "init".

jenswi-linaro commented on core/arch/arm/kernel/thread.c:
```diff
@@ -1344,34 +1363,88 @@ static void thread_rpc_alloc(size_t size, size_t align, unsigned int bt,
+			goto free_first;
+
+		nested_params = mobj_get_va(nested_mobj, page_offset);
+		if ((nested_params[0].attr & OPTEE_MSG_ATTR_TYPE_TMEM_OUTPUT)
```
OK, please add a `msg_param_attr_is_tmem()` function and use that here instead for clarity. A comment that this isn't a security check would also be helpful.

etienne-lms commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -928,6 +928,73 @@ static void set_pg_region(struct core_mmu_table_info *dir_info,
+	unsigned int idx;
+
+	for (i = 0; i < num_pages; i++) {
+		core_mmu_find_table(vaddr, UINT_MAX, &tbl_info);
```
if this fails, core should panic.
As this function can be called at runtime far after the inits, I doubt an `assert()` is sufficient.

etienne-lms commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -928,6 +928,73 @@ static void set_pg_region(struct core_mmu_table_info *dir_info,
+				goto out;
+			}
+
+			idx = core_mmu_va2idx(&tbl_info, vaddr);
```
This should check if the virtual page is already mapped.

etienne-lms commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -928,6 +928,73 @@ static void set_pg_region(struct core_mmu_table_info *dir_info,
+			}
+		}
+
+		core_mmu_set_entry(&tbl_info, idx, pages[i],
```
This should check that the virtual page is not already mapped. If already mapped:
- check attribute matches
- but it means there might be a balance issue at unmap time.

etienne-lms replied:
discussion moved to #1480.

etienne-lms commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -928,6 +928,73 @@ static void set_pg_region(struct core_mmu_table_info *dir_info,
+
+out:
+	if (ret != TEE_SUCCESS)
+		core_mmu_unmap(start, num_pages);
```
should unmap only the newly mapped memory.

lorc commented:
I have created separate PR for core_mmu_map/unmap functions: https://github.com/OP-TEE/optee_os/pull/1480

etienne-lms commented on core/arch/arm/kernel/user_ta.c:
```diff
@@ -645,11 +645,11 @@ static TEE_Result rpc_load(const TEE_UUID *uuid, struct shdr **ta,
 		return TEE_ERROR_OUT_OF_MEMORY;
 
-	*ta = phys_to_virt(phta, MEM_AREA_NSEC_SHM);
+	*ta = mobj_get_va(*mobj, 0);
```
This used to insure was SHM. Should you add `mobj_is_nonsec(*mobj);` or rely on `thread_rpc_alloc_payload()` to be safe? I guess the later is valid. So ok for this change. 
Then the case `(!*ta)` should either be asserted or panicked.

etienne-lms commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -643,7 +649,8 @@ bool core_pbuf_is(uint32_t attr, paddr_t pbuf, size_t len)
 	case CORE_MEM_NON_SEC:
-		return pbuf_is_inside(nsec_shared, pbuf, len);
+		return pbuf_is_inside(nsec_shared, pbuf, len) ||
+			pbuf_is_inside(ddr, pbuf, len);
```
This is the place where DDR layout should be checked.
Verifying in DDR is a good start but I think we must reject any SEC_RAM mapped DDR memories (if some platform declares in the DDR) and exclude also the SDP (bootime) registered memories (`pbuf_is_sdp_mem()`). And there are cases where core is inside DDR. This is about preventing core to faulty access a secure memory with non-secure attribute.

For this, we will need some devs. sorry. Maybe, before allowing SHM registration at boot time, we could allocate a table to list the DDR fragments eligible to dynamic SHM registration.

lorc replied:
In this case we will need support from `plat-` code. I think it can be look like
`register_nsec_ddr()` similar to `register_phys_mem()`.

And we will need this for every platform. 

OP-TEE will disable dynamic SHM and registered SHM feature in `tee_entry_exchange_capabilities()` if platform does not register nsec DDR. What do you think about this approach?

jenswi-linaro replied:
Sounds good to me.

etienne-lms replied:
nice, thanks.

lorc commented:
I have added new patch: https://github.com/OP-TEE/optee_os/pull/1232/commits/a514298b55bcb2fe3b6f5cda326010e92dc5d3d6 (`core_mmu: add non-secure DDR ranges support`).
Now platform code can define memory ranges. I altered `mobj` code to check those range during registered mobj  construction. This should ensure that OP-TEE will work only with non-secure memory.
Also I took into account all other comments.  Plus, I dropped patches for vexpress/qemu. I think, they are not relevant for this PR.

lorc commented:
Hello.

As #1480 was merged I rebased this PR onto the new master to remove reviewed patches.
What we will do next? I can create separate PR for https://github.com/OP-TEE/optee_os/pull/1232/commits/088ebf9fd14ea895c95470b6087e3779fc868bd7 (non-secure DDR ranges support) or we can continue review of whole patchset there.

etienne-lms commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -200,22 +200,44 @@ static struct tee_mmap_region *find_map_by_pa(unsigned long pa)
-extern const struct core_mmu_phys_mem __start_phys_sdp_mem_section;
-extern const struct core_mmu_phys_mem __end_phys_sdp_mem_section;
+extern const struct core_mmu_phys_mem __start_phys_nsec_ddr_section;
+extern const struct core_mmu_phys_mem __end_phys_nsec_ddr_section;
```
minor: move this right above `pbuf_is_nsec_ddr()`.

etienne-lms commented on core/arch/arm/include/mm/core_mmu.h:
```diff
@@ -464,6 +467,9 @@ static inline TEE_Result cache_op_outer(enum cache_op op __unused,
 bool cpu_mmu_enabled(void);
 
+/* Check if platform define nsec DDR range(s) */
+bool core_mmu_is_nsec_ddr_configured(void);
```
Renaming: I've understood `core_mmu_nsec_ddr_is_configured()` is better than the interrogative way. I even think `core_mmu_nsec_ddr_is_defined()` fits better.

jenswi-linaro commented on core/arch/arm/mm/mobj.c:
```diff
@@ -312,6 +308,293 @@ struct mobj *mobj_mm_alloc(struct mobj *mobj_parent, size_t size,
+	struct mobj_reg_shm *reg_shm =
+		to_mobj_reg_shm(mobj_mapped_shm->reg_shm);
+
+	offst += reg_shm->page_offset;
```
We generally try to avoid changing the arguments of a function.

jenswi-linaro commented on core/arch/arm/mm/mobj.c:
```diff
@@ -312,6 +308,293 @@ struct mobj *mobj_mm_alloc(struct mobj *mobj_parent, size_t size,
+
+	return &mobj_mapped_shm->mobj;
+err:
+	if (mobj_mapped_shm->reg_shm)
```
`mobj_free()` already checks for NULL

jenswi-linaro commented on core/arch/arm/mm/mobj.c:
```diff
@@ -312,6 +308,293 @@ struct mobj *mobj_mm_alloc(struct mobj *mobj_parent, size_t size,
+	if (mobj_mapped_shm->reg_shm)
+		mobj_free(mobj_mapped_shm->reg_shm);
+	if (mobj_mapped_shm->mm_entry)
+		tee_mm_free(mobj_mapped_shm->mm_entry);
```
`tee_mm_free()` already checks for NULL.

jenswi-linaro commented on core/arch/arm/mm/mobj.c:
```diff
@@ -312,6 +308,293 @@ struct mobj *mobj_mm_alloc(struct mobj *mobj_parent, size_t size,
+	vaddr_t pool_end;
+
+	core_mmu_get_mem_by_type(MEM_AREA_SHM_VASPACE, &pool_start, &pool_end);
+	if (pool_start == 0 || pool_end == 0)
```
`if (!pool_start || !pool_end)`

jenswi-linaro commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -1347,7 +1355,7 @@ static void check_pa_matches_va(void *va, paddr_t pa)
 #endif
 	if (!core_va2pa_helper(va, &p)) {
-		if (pa != p)
+		if (pa != p && p != 0)
```
`if (pa != p && p)`

jenswi-linaro commented on core/arch/arm/mm/mobj.c:
```diff
@@ -312,6 +308,293 @@ struct mobj *mobj_mm_alloc(struct mobj *mobj_parent, size_t size,
+		    TEE_MM_POOL_NO_FLAGS))
+		panic("Could not create shmem pool");
+
+	IMSG("Shared memory address range: %lx, %lx\n",
```
Please use `%" PRIxVA"` instead of `%lx`
Newline at the end not needed

jenswi-linaro commented on core/include/kernel/msg_param.h:
```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (c) 2017x, EPAM Systems
```
Please remove the "x"

jenswi-linaro commented on core/kernel/msg_param.c:
```diff
@@ -0,0 +1,217 @@
+{
+	uint32_t pages_cnt = 0;
+
+	if (params[num_params-1].attr & OPTEE_MSG_ATTR_FRAGMENT)
```
Please add a space before and after `-`

jenswi-linaro commented on core/kernel/msg_param.c:
```diff
@@ -0,0 +1,217 @@
+	if (params[num_params-1].attr & OPTEE_MSG_ATTR_FRAGMENT)
+		return -1;
+
+	for (uint32_t i = 0; i < num_params; i++) {
```
Please declare variables at the beginning of a block.

jenswi-linaro commented on core/kernel/msg_param.c:
```diff
@@ -0,0 +1,217 @@
+ *   <0 on error
+ */
+int msg_param_extract_pages(struct optee_msg_param *params, paddr_t *pages,
+			    uint32_t num_params)
```
I'd prefer if `size_t` was used instead of `uint32_t` in this function and also that the function returned an `ssize_t`
Fixed size integers should only be used when it's important that just that size is used.

lorc replied:
Actually, it is preferred to use simple integral types in such cases. So it would be better to use `unsigned int` instead of `uint32_t`. But, as there are no coding standard for OP-TEE defined, I'll do as you say.

jenswi-linaro replied:
Yes, `unsigned int` would have been OK also. But it is a bit tedious to write and there's actually very few cases where it's a better choice that a `size_t`.
We do have a coding standard, it's same as Linux kernel style, with one or two exceptions.

jenswi-linaro commented on core/kernel/msg_param.c:
```diff
@@ -0,0 +1,217 @@
+ *  struct shmem_mapping of mapped buffer on success
+ *  NULL on error.
+ */
+struct mobj *msg_param_map_buffer(paddr_t pa_params, uint32_t num_params)
```
Please use `size_t` instead of `uint32_t` in this function.

jenswi-linaro commented on core/kernel/msg_param.c:
```diff
@@ -0,0 +1,217 @@
+	size_t args_size;
+
+	args_size = OPTEE_MSG_GET_ARG_SIZE(num_params);
+	max_pages = args_size/SMALL_PAGE_SIZE + 1;
```
A space before and after that `/`

jenswi-linaro commented on core/kernel/msg_param.c:
```diff
@@ -0,0 +1,217 @@
+	struct mobj *mobj = NULL;
+	struct optee_msg_param *params;
+	paddr_t *pages = NULL;
+	uint32_t max_pages, pages_cnt = 1;
```
We generally declare one variable per line.

jenswi-linaro commented on core/kernel/msg_param.c:
```diff
@@ -0,0 +1,217 @@
+
+	params = mobj_get_va(mobj, pa_params & SMALL_PAGE_MASK);
+
+	for (uint32_t i = 0; i < num_params; i++) {
```
Declare variables at the start of a block.

jenswi-linaro commented on core/arch/arm/kernel/thread.c:
```diff
@@ -1344,34 +1357,87 @@ static void thread_rpc_alloc(size_t size, size_t align, unsigned int bt,
-	if (params[0].attr != OPTEE_MSG_ATTR_TYPE_TMEM_OUTPUT)
-		goto fail;
+	if (params[0].attr == OPTEE_MSG_ATTR_TYPE_NESTED) {
+		paddr_t *pages;
```
This entire block should probably go into a helper function instead.

jenswi-linaro commented on core/arch/arm/kernel/thread.c:
```diff
@@ -1344,34 +1357,87 @@ static void thread_rpc_alloc(size_t size, size_t align, unsigned int bt,
+		goto free_first;
+
+	if (!mobj_is_nonsec(mobj))
+		panic("This should not happen");
```
Could we have `assert(mobj_is_nonsec(mobj))` instead?

jenswi-linaro commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -104,6 +104,8 @@ register_phys_mem(MEM_AREA_TEE_RAM, CFG_TEE_RAM_START, CFG_TEE_RAM_PH_SIZE);
 register_phys_mem(MEM_AREA_TA_RAM, CFG_TA_RAM_START, CFG_TA_RAM_SIZE);
 register_phys_mem(MEM_AREA_NSEC_SHM, CFG_SHMEM_START, CFG_SHMEM_SIZE);
 
+static bool arm_va2pa_helper(void *va, paddr_t *pa);
```
Is this still needed?

jenswi-linaro commented on core/arch/arm/tee/entry_fast.c:
```diff
@@ -102,7 +102,19 @@ static void tee_entry_exchange_capabilities(struct thread_smc_args *args)
+	} else {
+		EMSG("No NS DDR configured for the platform. Disabling dynamic SHM");
+	}
+#else
```
We wouldn't need this `#else` if the `#ifdef CFG_SMALL_PAGE_USER_TA` above was moved down one line.

jenswi-linaro commented on core/arch/arm/tee/entry_std.c:
```diff
@@ -92,6 +93,26 @@ static TEE_Result assign_mobj_to_param_mem(const paddr_t pa, const size_t sz,
+	if (mobj_get_pa(rmem_mobj, 0, 0, &b) != TEE_SUCCESS)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (rmem_mobj == NULL)
```
`mobj_get_pa()` already checks the supplied mobj for NULL.

jenswi-linaro commented on core/arch/arm/tee/entry_std.c:
```diff
@@ -350,21 +514,31 @@ void tee_entry_std(struct thread_smc_args *smc_args)
+	arg = mobj_get_va(mobj, 0);
+	assert(arg && mobj_is_nonsec(mobj));
+
+	num_params = arg->num_params;
```
We should either use the value previously read by `get_cmd_buffer()` or `map_cmd_buffer()`, or we'll need to check that the value we read now is small enough for the `mobj`. I'd prefer to only read once to keep it simple.

jenswi-linaro commented on core/tee/tee_fs_rpc_cache.c:
```diff
@@ -28,27 +28,29 @@
 	void *va;
 
-	if (!size)
+	if (!size || !mobj)
```
`mobj` shouldn't be NULL, just as `cookie` isn't supposed to be NULL.

jenswi-linaro commented on core/tee/tee_rpmb_fs.c:
```diff
@@ -458,15 +461,19 @@ static TEE_Result tee_rpmb_alloc(size_t req_size, size_t resp_size,
-	if (!mem->phreq || !mem->phresp) {
+
+	mem->phreq_mobj =
+		thread_rpc_alloc_payload(req_s, &mem->phreq_cookie);
```
This line could be folded up.

etienne-lms commented on core/arch/arm/mm/mobj.c:
```diff
@@ -155,12 +155,6 @@ struct mobj *mobj_phys_alloc(paddr_t pa, size_t size, uint32_t cattr,
-		DMSG("Expect %#x alignment", CORE_MMU_USER_PARAM_SIZE);
-		return NULL;
-	}
-
```
why removing this ?

lorc replied:
Because I use mobj_phys to represent shared buffers. They are not always page-aligned.

etienne-lms commented on core/arch/arm/kernel/thread.c:
```diff
@@ -1186,26 +1181,39 @@ void thread_rpc_free_arg(uint64_t cookie)
-	*arg = pa;
+	if (!ALIGNMENT_IS_OK(pa, struct optee_msg_arg))
+		goto err;
+	/* Check if this region is in static shared space */
```
The test below does not check only the static mapping. `core_pbuf_is()`->`pbuf_is_nsec_ddr()` will return true for `size`@`pa` inside the area defined by table `__start/end_phys_nsec_ddr_section`.

lorc replied:
Hmmm, why? As I can see, `core_pbuf_is(CORE_MEM_NSEC_SHM)` calls `pbuf_inside_map_area(pbuf, len, map_nsec_shm)` and `map_nsec_shm` holds only static SHM buffer.

lorc replied:
Thanks. Good idea. This fixed regression caused by v2 patch. @jenswi-linaro stepped right into it.

etienne-lms commented on core/arch/arm/mm/mobj.c:
```diff
@@ -312,6 +306,291 @@ struct mobj *mobj_mm_alloc(struct mobj *mobj_parent, size_t size,
+{
+	assert(mobj->ops == &mobj_mapped_shm_ops);
+
+	return attr == CORE_MEM_NON_SEC || attr == CORE_MEM_EXTRAM ||
```
`CORE_MEM_EXTRAM` cover the physical DDR. It may cover secure memory regions.
Can't we stick to `attr == CORE_MEM_NON_SEC || attr == CORE_MEM_NSEC_SHM`?

etienne-lms commented on core/kernel/msg_param.c:
```diff
@@ -0,0 +1,221 @@
+			     size_t offset, size_t size,
+			     uint64_t cookie, enum msg_param_mem_dir dir)
+{
+	if (mobj_matches(mobj, CORE_MEM_EXTRAM)) {
```
EXTRAM covers the whole DDR ([core_mmu.c#L731](https://github.com/OP-TEE/optee_os/blob/master/core/arch/arm/mm/core_mmu.c#L731)), eventually including secure memory areas.

lorc commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -506,7 +526,15 @@ static void init_mem_map(struct tee_mmap_region *memory_map, size_t num_elems)
+
+	verify_special_mem_areas(memory_map, num_elems,
+				 &__start_phys_nsec_ddr_section,
+				 &__end_phys_nsec_ddr_section, "NSEC DDR");
```
You propose to force disable nsec ddr regions when SDP is enabled?
I have created function that cross-checks SDP and NSEC DDR regions. I thing it is better approach.

lorc replied:
Actually, we want to bother with SDP. It is one of the use cases we want to support here at EPAM.

jenswi-linaro commented on core/arch/arm/kernel/thread.c:
```diff
@@ -1307,15 +1318,65 @@ static void thread_rpc_free(unsigned int bt, uint64_t cookie)
+
+	nested_params = mobj_get_va(nested_mobj, page_offset);
+	if (!msg_param_attr_is_tmem(nested_params)) {
+		mobj_free(nested_mobj);
```
Remove, causes double free below `out:`
Fix the other `goto out` cases below too.

jenswi-linaro commented on core/kernel/msg_param.c:
```diff
@@ -0,0 +1,221 @@
+			     size_t offset, size_t size,
+			     uint64_t cookie, enum msg_param_mem_dir dir)
+{
+	if (mobj_matches(mobj, CORE_MEM_NON_SEC)) {
```
I was testing this pull request without any changes in the other gits. I was expecting everything to just work falling back to using the predefined shm pool, especially since this log entry:
```
ERROR:   TEE-CORE:tee_entry_exchange_capabilities:113: No NS DDR configured for the platform. Disabling dynamic SHM
```
But it seems this test still causes the param to be initialized with registered SHM causing a failure later.

lorc replied:
Oops. This regression was introduced in v2. Should be fixed in v4.
To test this PR, you will need this two patches: https://github.com/lorc/optee_os/commit/c85b463e7f0dc7c0e0aa43510fe0f91ee17c1345 ,
https://github.com/lorc/optee_os/commit/cfccc287cca873b22491b0d54ceed3d4bd236629
I decided not to include them in this PR.

jenswi-linaro replied:
For the record, with v4 the code seems to fall back to the shm pool as expected and xtest is all OK.

etienne-lms commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -200,107 +200,155 @@ static struct tee_mmap_region *find_map_by_pa(unsigned long pa)
+	const struct core_mmu_phys_mem *sdp;
+	const struct core_mmu_phys_mem *nsec_ddr;
+
+	if (sdp_start == sdp_end || ddr_start == ddr_end)
```
minor: test on `ddr_start == ddr_end` is sufficient to prevent a useless scan. Could even remove the full test.

lorc replied:
I like symmetry in this test, so I'd prefer to leave it as is, if you don't mind. Imagine, that someone will switch inner loop with outer later.

jenswi-linaro commented:
It seems I failed to answer this question by @lorc from  Apr 7:
> @jenswi-linaro, If I remember well, I saw that approach with array of params in your experiments with registered memory. But we can pass just plain array of page addresses if you think that it is will be better.

Yes, plain array of page addresses should be much more efficient. I think it would make the code easier too understand.

By the way, I've tested this PR in the current state successfully and it looks very promising.

So far I think we are OK with the first two commits `"core_mmu: add non-secure DDR ranges support"` and `"mobj: add mobj_reg_shm and mobj_mapped_shm"`, perhaps those could be squashed and merged via a new PR to shrink this PR further. With those commits in place I can add code to get non-secure DDR range from DT instead of (or as a complement to) hardcoding it in the QEMU platforms.

The reason I'm so cautious with taking the complete PR in the current state is that this PR is an ABI change so we need to get the ABI right. Then there's also the changes in the kernel driver, perhaps it turns out that the ABI could be changed in some way to make it easier in the driver.

lorc commented:
@jenswi-linaro, Okay, I have created separate PR #1527. In the meantime I'll rework ABI. 
Just to be sure: plain old array of physical page addresses stored in `uint64_t` (even on 32-bit platforms without LPAE), right?

jenswi-linaro commented:
@lorc, yes sounds good. Thanks for the patience/persistence.

lorc commented:
@jenswi-linaro, one more question about this ABI. I'm going to store list of page in the same registered buffer. So, when app calls `TEE_RegisterSHM(buf)`, kernel driver will fill list of the pages into the `buf`. Pointer to the first page of `buf` will be passed in a command buffer.
There are two consequences: 
 - kernel will overwrite data in app's buffer
 - app can find list of the pages in the registered buffer

GP API specification say nothing about such behavior. How do you think: is this acceptable?

etienne-lms commented:
Wouldn't it be more easy if registering SHM reference was using a buffer from the default physically contiguous SHM negotiated at OP-TEE init ?

jenswi-linaro commented:
I have another idea. As you pointed out we can't use the user space pages to carry the physical pointers. We scrap `OPTEE_MSG_ATTR_FRAGMENT` and add a `OPTEE_MSG_ATTR_NONCONTIG` or similar. 

What this `OPTEE_MSG_ATTR_NONCONTIG` means is that the physical address in `struct optee_msg_param_tmem:but_ptr` points to a page holding a list of physical addresses, each address is one "user space" page. The last address in this page holds the address of next page of addresses if needed. `struct optee_msg_param_tmem:size` is still the size in bytes (normally page aligned, except in the corner case below) of the shared memory, from this it's easy to determine how many pages are used to hold the pointers etc. When `OPTEE_MSG_ATTR_NONCONTIG` is used then the shared memory is always page aligned, eventual offsets into the first page etc has to be taken care of in an outer layer.

This model isn't too complicated it's the same as what's in use today except that with `OPTEE_MSG_ATTR_NONCONTIG` in combination with `OPTEE_MSG_ATTR_TYPE_TMEM_*` what's described above is activated. This should work both for RPC allocations (`OPTEE_MSG_RPC_CMD_SHM_ALLOC`) and `OPTEE_MSG_CMD_REGISTER_SHM`.

I think it's important (especially for the virtualization use case) to be able to get rid shared memory pool we're using today completely. We can still support the shared memory pool, but it's not a requirement.

So how can we deal with a normal invoke with `OPTEE_MSG_ATTR_TYPE_TMEM_*` in combination with `OPTEE_MSG_ATTR_NONCONTIG` where the shared memory buffer isn't page aligned? One way could be to let the `buf_ptr & SMALL_PAGE_MASK` bits signify the offset into the shared memory buffer while `buf_ptr & ~SMALL_PAGE_MASK` is the pointer to the physical page holding the list of physical addresses.

Does it make sense? Can it be simplified further? Am I missing any corner case?

lorc commented:
@jenswi-linaro, thanks for the idea. I'm currently implementing it. Do deal with non-page aligned shared buffer I prefer to store beginning of the shared buffer as the first element of the array. So `pages_array[0] & ~SMALL_PAGE_MASK` will point to the firts page and `pages_array[0]` will point to the beginning of the buffer. 

But I see another problem there: RPC calls. It is easy to allocate buffer for pages array in `OPTEE_MSG_CMD_REGISTER_SHM` because this is synchronous call and I can free this buffer right after SMC invocation. 
In RPC case, I will allocate buffer for user pages array in response to  `OPTEE_MSG_RPC_CMD_SHM_ALLOC`. But when I should free it? It should be deferred until normal return from SMC. My previous implementation used allocated buffer to store pages, because in this case buffer was provided either by driver itself or by supplicant. And it was okey, because RPC code didn't allocated any new buffer. What do you think? Should I use the same approach, or it is better to develop some generic mechanism for deferred cleanup?

jenswi-linaro commented:
It's not so important where the offset is stored, but I think at least one special case can be removed if it's coded into the lower bits of `buf_ptr`. I'd rather have registered shared memory page aligned as that's how it's mapped nonetheless. Then the offset isn't part of the shared memory object, instead it's part of the parameter. We should be careful to avoid adding features that aren't strictly needed since those usually brings special cases that has to be dealt with.

Regarding the buffer for the pages array for RPC:
We should allocate this buffer just as for the invoke cases, we just need to figure how when we can free it.
We could store it somewhere in `optee_do_call_with_arg()`, but as we can't know for sure when the information has been used we can't free it until the function returns (or the shared memory object is freed). Another option is to store it in `struct tee_shm`, but then we'll have the problem storing this information in a generic way, we can't have OP-TEE specific stuff in that struct.

lorc commented:
Okay, so I decided to do as you say and store offset in `buf_ptr`. But to do this, I had to ensure that `buf_ptr` always points to a page boundary. I made necessary adjustments on the kernel side and now it is fine.

So, there are v5 delta. To produce #1527  I had to squash v1-v4 deltas and rebase to the master branch. 
I don't ask you to review whole PR, because it will be rebased again after #1527 merge. But can you please take a look at v5 delta (https://github.com/OP-TEE/optee_os/pull/1232/commits/5681beb9f6886f1806f2d949a56a70ad26106b48) to confirm new ABI implementation?

Patch for the kernel is on the way.

jenswi-linaro commented:
Looks good. :-) I've added a few comments in the "v5 delta".

Only misunderstanding is that I'd like `OPTEE_MSG_ATTR_TYPE_NONCONTIG` to be a separate bit that is used together with `OPTEE_MSG_ATTR_TYPE_TMEM_*` and only `OPTEE_MSG_ATTR_TYPE_TMEM_*`. Without `OPTEE_MSG_ATTR_TYPE_TMEM_*`  `OPTEE_MSG_ATTR_TYPE_NONCONTIG` doesn't mean anything.

`copy_in_params()` needs to deal with the `OPTEE_MSG_ATTR_TYPE_NONCONTIG` bit also where memory hasn't been registered in advance. We could of course wait with this until it's a problem that needs to be optimized, but I think this is such a small change compared to everything that it's better to do it now than having to add yet another capability later when it's introduced.

lorc commented:
@jenswi-linaro, thanks for the review. I had pushed v6 delta.

I want to clarify `copy_in_params()` with `OPTEE_MSG_ATTR_TYPE_TMEM_*` use case. You want to pass `TMEM_*` buffers without prior registration, right? That would save use two OP-TEE calls, but will require some rework on NW side. Anyways, I'll try to implement this.

jenswi-linaro commented:
Yes, `copy_in_params()` should be able to with with `OPTEE_MSG_ATTR_TYPE_TMEM_*` where the `OPTEE_MSG_ATTR_NONCONTIG` bit can be both set or clear. The latter case to preserve the ABI and be backwards compatible. The former case as an optimization to avoid three calls just to make one invoke with unregistered physically non-contiguous memory.

lorc commented:
@jenswi-linaro, thanks for the clarification.
During implementing and testing this feature I encountered two issues:

 * First one is backward parameters passing (when returning results from TA back to client app). Without `OPTEE_MSG_ATTR_NONCONTIG` we can easily calculate buffer offset: `p->u.memref.shm_offs = mp->u.tmem.buf_ptr - shm_pa`. 
 But what to do with `OPTEE_MSG_ATTR_NONCONTIG` ?  `mp->u.tmem.buf_ptr` holds pointer to the page list. Probably we can calculate offset on OP-TEE side and put it into `mp->u.tmem.buf_ptr`. But this will be semantically incorrect. Probably we can put real PA into `mp->u.tmem.buf_ptr` on OP-TEE side. But then kernel driver should perform PA->VA translation for userspace code. 

 * Another issue is with userspace mapping. As you remember, you have used `get_user_pages_fast()` to get client pages. Issue is in with this line in `regression_1000.c:145`:
 `	static const uint8_t sha256_in[] = { 'a', 'b', 'c' };`  
 As you can see, this is very short static const array. Compiler have put it in `.text` section. Problem is that `get_user_pages()` can't get reference to pages in this section, because technically it is not a part of application's memory, it is mmaped from executable file. 
 I don't know if memory reference to `.text` section can be considered as a programmer error, because this memory chunk serves as input parameter, so it can reside in RO memory. What do you thing about this issue?
(BTW, output buffer is also declared as `static const` and this is clearly programmer error).

jenswi-linaro commented:
Issue 1:
The pointer value isn't supposed to be updated so there should be any need to update that on the way back.

Issue 2:
Hmm, we'll need to do some magic. On error perhaps we can let the the client lib check the arguments and use temporary buffers for stuff that's in read-only pages. How the lib reliably can tell if a page is read-only mapped or not I don't know though.

lorc commented:
Hi @jenswi-linaro, 
Sorry for delay. It took some time to test this feature. So there are v7 delta patch that implements it. I think, now it is quite fine.
It will require some rework in ioctl interface, but let's discuss that in driver PR.

As for that two issues:

Issue 1: I also think that pointer value should not be updated, but optee driver does this for some reason.
Issue 2: I think that simple fallback to shadow buffer will be enough.

jenswi-linaro commented:
Thanks @lorc 
Issue 1: I don't know why either, shouldn't be needed though. Note that `params_to_user()` in `drivers/tee/tee_core.c` doesn't.

Issue 2: sounds good.

I'll not be able to do much reviewing this week, but next week I'll prioritize this.

lorc commented:
Hello. So, after #1527 was merged, I rebased this PR onto master branch. 
Also I added two new patches:
 https://github.com/OP-TEE/optee_os/pull/1232/commits/5e668b512000c13d58bf19abdebc1f10c77025e9  adds new `mobj` type instead of `mobj_phys` which I used earlier.
 https://github.com/OP-TEE/optee_os/pull/1232/commits/3d4b5d7f3ff5ee6aafc35743573fcc36ca09efa2 fixes another dumb bug in newly added `mobj_reg_shm`

lorc commented:
Okay, I created #1582. It includes three patches that I took from this PR.

lorc commented:
During review this PR evolved into something completely different, multiple PRs were detached from it. Current code base does not resemble what was pushed originally. So I decided to close it and push new patches for registered shm in new PRs. 
Thank you for your time, guys.
